### PR TITLE
add openvpn_connected_clients

### PR DESCRIPTION
**Pull request by Sispheor**
- add openvpn_connected_clients gauge
- add a docker file to build and/or use the exporter
- minor code review following [golang recommandation](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)
- switch from deprecated prometheus http handler to the new one

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# docker build --force-rm=true -t openvpn_exporter .
+
+# RUN:
+# docker run -it -p 9176:9176 -v /path/to/openvpn_server.status:/etc/openvpn_exporter/server.status openvpn_exporter
+
+FROM golang as builder
+
+# Grenoble testing env
+ENV http_proxy="http://web-proxy.gre.hpecorp.net:8080" \
+    https_proxy="http://web-proxy.gre.hpecorp.net:8080"
```
Comment (EdSchouten) [inline Dockerfile]: This seems like some left-over from development, right?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+RUN mkdir /app
+RUN mkdir /go/src/app
+ADD . /go/src/app
+WORKDIR /go/src/app
+
+# Go dep
+RUN go get -u github.com/prometheus/client_golang/prometheus && \
+    go get -u github.com/prometheus/client_golang/prometheus/promhttp
```
Comment (EdSchouten) [inline Dockerfile]: Instead of listing all dependencies explicitly, we could use `go get -d ./...` here.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+# Grenoble testing env
+ENV http_proxy="http://web-proxy.gre.hpecorp.net:8080" \
+    https_proxy="http://web-proxy.gre.hpecorp.net:8080"
+
+# add compiled binary
+COPY --from=builder /app/openvpn_exporter /openvpn_exporter
+
+# add a default file to be processed
+ADD examples/server2.status /etc/openvpn_exporter/server.status
```
Comment (EdSchouten) [inline Dockerfile]: These examples are just included in the repository to test what the exporter does. There is no need for including these in any container image.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        -ldflags '-extldflags "-static"' && \
+  ls
+
+# Create the second stage with a basic image.
+# this will drop any previous
+# stages (defined as `FROM <some_image> as <some_name>`)
+# allowing us to start with a fat build image and end up with
+# a very small runtime image.
+
+FROM busybox
```
Comment (EdSchouten) [inline Dockerfile]: This container image is now incredibly light-weight. What about using `FROM scratch`?

File: README.md [hunk 5]
```diff
@@ -48,4 +50,49 @@ openvpn_server_client_sent_bytes_total{common_name="...",connection_time="...",r
...
+
+To use with docker you must mount your status file to `/etc/openvpn_exporter/server.status`.
+```
+docker run -it -p 9176:9176 -v /path/to/openvpn_server.status:/etc/openvpn_exporter/server.status openvpn_exporter
+```
+
+Metrics should be available on your host IP: http://<host_ip>:9176/metrics. E.g: http://10.39.9.94:9176/metrics
+
+
+## Get a standalone executable binary
```
Comment (EdSchouten) [inline README.md]: We already had `build_static.sh` for this purpose. Should this script be reimplemented on top of this Dockerfile now? Maybe also worth mentioning it here in the documentation.

File: openvpn_exporter.go [hunk 6]
```diff
@@ -52,8 +53,12 @@ var (
 		[]string{"status_path"}, nil)
 
 	// Metrics specific to OpenVPN servers.
+	openvpnConnectedClientsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("openvpn", "", "connected_clients"),
+		"Number Of Connected Clients", nil, nil)
```
Comment (EdSchouten) [inline openvpn_exporter.go]: Is there any need for having this as a metric explicitly? One could just use `count()` for this, right?

File: Dockerfile (hunk 3 above)
Comment (Sispheor) [reply Dockerfile]: This is just to allow the image to run without any error if the user doesn't provide a file with a volume. As the `-openvpn.status_paths` is mandatory. I can remote it anyway if you prefer.

File: openvpn_exporter.go (hunk 6 above)
Comment (Sispheor) [reply openvpn_exporter.go]: This is actually the only metric I use. I thought it would be nice to have it directly.

File: Dockerfile (hunk 4 above)
Comment (Sispheor) [reply Dockerfile]: You are right, I used busybox for dev and I forgot to switch back to scratch.

File: Dockerfile (hunk 2 above)
Comment (Sispheor) [reply Dockerfile]: Right, I didn't know this command. I'll update it. Thanks

File: Dockerfile (hunk 4 above)
Comment (Sispheor) [reply Dockerfile]: Oh yes, I remember now why I finally let busybox. It's to have the "cp" command that I use to copy the compiled binary like mentioned in the doc.
What do you prefer then? I switch to `scratch` anyway?

File: Dockerfile (hunk 1 above)
Comment (Sispheor) [reply Dockerfile]: Yes I removed it already.

File: README.md (hunk 5 above)
Comment (Sispheor) [reply README.md]: Yes I saw that. I left it.
Here we can:
- delete the sh script and so redirect users to the docker run command I've added to get a static file
- doc the sh script and remove the part about getting a static file from docker command

Up to you.

File: Dockerfile (hunk 4 above)
Comment (EdSchouten) [reply Dockerfile]: It looks like `docker save` can be used to export files from an image. We could use that in combination with a `FROM scratch` container.

Comment (Sispheor) [reply Dockerfile]: It seems that the generated file can be then only used by `docker load` command. Maybe I miss something.

Comment (EdSchouten) [reply Dockerfile]: Oh, hmmm... Indeed. That's annoying. Well, just stick to Busybox then. :-)

File: README.md (hunk 5 above)
Comment (EdSchouten) [reply README.md]: Deleting the shell script sounds fine by me, as long as it's documented.

File: Dockerfile (hunk 4 above)
Comment (Sispheor) [reply Dockerfile]: Okey ;)

File: README.md (hunk 5 above)
Comment (Sispheor) [reply README.md]: Done

**Comment by Sispheor** [comment]
I've updated following all your recommandation. Is everything ok for you now @EdSchouten ?

File: openvpn_exporter.go [hunk 7]
```diff
@@ -52,8 +53,12 @@ var (
 		[]string{"status_path"}, nil)
 
 	// Metrics specific to OpenVPN servers.
+	openvpnConnectedClientsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("openvpn", "", "connected_clients"),
```
[the remark this answers is not in the document]
Comment (Sispheor) [reply openvpn_exporter.go]: Done

**Pull request merged**